custom/icds_reports/data_pull/data_pulls.py

```diff
@@ -193,3 +200,75 @@ def _dump_consolidated_data(result):
             row['State'] = state_name
             writer.writerow(row)
         return result_file
+
+
+class VHSNDMonthlyReport(MonthBasedDataPull):
+    slug = "vhsnd_monthly_report"
+    name = "VHSND monthly report"
+    queries = [
+        VHSNDMonthlyCount
+    ]
+
+    def post_run(self, data_files):
+        state_results = self._consolidate_data(data_files)
+        return self._dump_consolidated_data(state_results)
+
+    def _consolidate_data(self, data_files):
+        result = defaultdict(dict)
+        for filename, filestream in data_files.items():
+            filestream.seek(0)
+            reader = csv.DictReader(filestream)
+            for row in reader:
+                state_name = row['state_name']
+                vhsnd_date = parse_date(row['vhsnd_date_past_month']).strftime('%d/%m/%Y')
+                data_key = (row['state_name'], row['district_name'], row['block_name'],
+                            row['supervisor_name'], row['awc_name'])
+                if data_key in result[state_name]:
+                    result[state_name][data_key].append(vhsnd_date)
+                else:
+                    result[state_name][data_key] = [vhsnd_date]
+
+        return self._format_consolidated_data(result)
+
+    def _format_consolidated_data(self, result):
+        state_results = defaultdict(dict)
+        self.month_date = datetime.datetime.strptime(self.month, '%Y-%m-%d')
+        num_days = calendar.monthrange(self.month_date.year, self.month_date.month)[1]
+        days = [datetime.date(self.month_date.year, self.month_date.month, day) for day in range(1, num_days + 1)]
+
+        dates = [day.strftime('%d/%m/%Y') for day in days]
+
+        headers = ['State', 'District', 'Block', 'Sector', 'AWC']
+        headers.extend(dates)
+        headers.append('Grand Total')
```

Comment:
there is a lot happening in this method,
can we simply `extract _get_headers` and may be `setup_state_results_filestream`?

Reply:
refactored.please check